### PR TITLE
Bug fixes

### PR DESCRIPTION
**Pull request by iamshital**
1. Fixed issue, where windows was failing to access the files if file path is more than 255 characters.
2. Added "ppa" kernel installation support to CustomKernelInstall.sh [Packages in “PPA for Canonical Kernel Team”](https://launchpad.net/~canonical-kernel-team/+archive/ubuntu/ppa/+packages)
3. Fixed issue of duplicate test name in test summary
4. Split VERIFY-DEPLOYMENT-PROVISION stage in four subtests -
Example: 
VERIFY-DEPLOYMENT-PROVISION Execution Time: 9.92 minutes
	VERIFY-DEPLOYMENT-PROVISION : PASS  
		  FirstBoot : PASS 
		  FirstBoot : Call Trace Verification : PASS 
		  Reboot : PASS 
		  Reboot : Call Trace Verification : PASS 
5. Fixed Telemetry column name
6. Added VERIFY-DEPLOYMENT-PROVISION-SRIOV test

### CODE REVIEW

File: Testscripts/Linux/customKernelInstall.sh [hunk 1]
```diff
@@ -122,6 +122,54 @@ InstallKernel()
                         LogMsg "CUSTOM_KERNEL_SUCCESS"
                         UpdateTestState $ICA_TESTCOMPLETED
                 fi
+        elif [ "${CustomKernel}" == "proposed" ]; then
+                DISTRO=`grep -ihs "buntu\|Suse\|Fedora\|Debian\|CentOS\|Red Hat Enterprise Linux" /etc/{issue,*release,*version}`
```
Comment (ader1990) [inline Testscripts/Linux/customKernelInstall.sh]: put this in a method and reuse it (duplicate code)

File: Testscripts/Linux/customKernelInstall.sh [hunk 2]
```diff
@@ -122,6 +122,54 @@ InstallKernel()
...
+                if [[ $DISTRO =~ "Xenial" ]];
+                then
+                        LogMsg "Enabling proposed repositry..."
+                        echo "deb http://archive.ubuntu.com/ubuntu/ xenial-proposed restricted main multiverse universe" >> /etc/apt/sources.list
+                        rm -rf /etc/apt/preferences.d/proposed-updates
```
Comment (ader1990) [inline Testscripts/Linux/customKernelInstall.sh]: always use quotes when working with strings (this applies for all current and future code)

File: Testscripts/Linux/customKernelInstall.sh [hunk 3]
```diff
@@ -122,6 +122,54 @@ InstallKernel()
...
+                        echo "deb http://archive.ubuntu.com/ubuntu/ trusty-proposed restricted main multiverse universe" >> /etc/apt/sources.list
+                        rm -rf /etc/apt/preferences.d/proposed-updates
+                        LogMsg "Installing linux-image-generic from proposed repository."
+                        apt -y update >> $logFolder/build-CustomKernel.txt 2>&1
+                        apt -y --fix-missing upgrade >> $logFolder/build-CustomKernel.txt 2>&1
+                        kernelInstallStatus=$?
+                fi
+                UpdateTestState $ICA_TESTCOMPLETED
+                if [ $kernelInstallStatus -ne 0 ]; then
+                        LogMsg "CUSTOM_KERNEL_FAIL"
```
Comment (ader1990) [inline Testscripts/Linux/customKernelInstall.sh]: keep the indentation consistent (never use tabs or more or less than 4 spaces)

File: RunTests.ps1 [hunk 4]
```diff
@@ -67,7 +67,26 @@ Get-ChildItem .\Libraries -Recurse | Where-Object { $_.FullName.EndsWith(".psm1"
 try 
 {
     #region Prepare / Clean the powershell console.
-    $WorkingDirectory = Split-Path -parent $MyInvocation.MyCommand.Definition 
+    $MaxDirLength = 32
+    $WorkingDirectory = Split-Path -parent $MyInvocation.MyCommand.Definition
+    Set-Variable -Name shortRandomNumber -Value $(Get-Random -Maximum 99999 -Minimum 11111) -Scope Global
```
Comment (ader1990) [inline RunTests.ps1]: create a function for the random number logic

File: RunTests.ps1 [hunk 5]
```diff
@@ -67,7 +67,26 @@ Get-ChildItem .\Libraries -Recurse | Where-Object { $_.FullName.EndsWith(".psm1"
...
+    Set-Variable -Name shortRandomWord -Value $(-join ((65..90) | Get-Random -Count 4 | ForEach-Object {[char]$_})) -Scope Global
```
Comment (ader1990) [inline RunTests.ps1]: create a function for the random word logic

File: RunTests.ps1 [hunk 6]
```diff
@@ -67,7 +67,26 @@ Get-ChildItem .\Libraries -Recurse | Where-Object { $_.FullName.EndsWith(".psm1"
...
+        Write-Host "Current working directory '$WorkingDirectory' length is greather than $MaxDirLength. Need to change the working directory."
+        $TempWorkspace = "$(Split-Path $OriginalWorkingDirectory -Qualifier)"
+        New-Item -ItemType Directory -Path "$TempWorkspace\LISAv2" -Force -ErrorAction SilentlyContinue | Out-Null
+        New-Item -ItemType Directory -Path "$TempWorkspace\LISAv2\$shortRandomWord$shortRandomNumber" -Force -ErrorAction SilentlyContinue | Out-Null
+        $finalWorkingDirectory = "$TempWorkspace\LISAv2\$shortRandomWord$shortRandomNumber"
+        $tmpSource = '\\?\' + "$OriginalWorkingDirectory\*"
+        Write-Host "Copying current workspace to $finalWorkingDirectory"
+        Copy-Item -Path $tmpSource -Destination $finalWorkingDirectory -Recurse -Force -ErrorAction SilentlyContinue| Out-Null
+        Set-Location -Path $finalWorkingDirectory | Out-Null
+        Write-Host "Wroking directory changed to $finalWorkingDirectory"
```
Comment (ader1990) [inline RunTests.ps1]: typo

File: RunTests.ps1 [hunk 7]
```diff
@@ -67,7 +67,26 @@ Get-ChildItem .\Libraries -Recurse | Where-Object { $_.FullName.EndsWith(".psm1"
...
+    Set-Variable -Name shortRandomNumber -Value $(Get-Random -Maximum 99999 -Minimum 11111) -Scope Global
+    Set-Variable -Name shortRandomWord -Value $(-join ((65..90) | Get-Random -Count 4 | ForEach-Object {[char]$_})) -Scope Global
+    if ( $WorkingDirectory.Length -gt $MaxDirLength)
+    {
+        $OriginalWorkingDirectory = $WorkingDirectory
+        Write-Host "Current working directory '$WorkingDirectory' length is greather than $MaxDirLength. Need to change the working directory."
+        $TempWorkspace = "$(Split-Path $OriginalWorkingDirectory -Qualifier)"
+        New-Item -ItemType Directory -Path "$TempWorkspace\LISAv2" -Force -ErrorAction SilentlyContinue | Out-Null
+        New-Item -ItemType Directory -Path "$TempWorkspace\LISAv2\$shortRandomWord$shortRandomNumber" -Force -ErrorAction SilentlyContinue | Out-Null
+        $finalWorkingDirectory = "$TempWorkspace\LISAv2\$shortRandomWord$shortRandomNumber"
```
Comment (ader1990) [inline RunTests.ps1]: never ever hardcode the repo name inside the repo code, it is a huge problem when you want to fork a repo as the repo name might change at fork time and you end up with an unusable code

File: RunTests.ps1 [hunk 8]
```diff
@@ -67,7 +67,26 @@ Get-ChildItem .\Libraries -Recurse | Where-Object { $_.FullName.EndsWith(".psm1"
...
+        $tmpSource = '\\?\' + "$OriginalWorkingDirectory\*"
+        Write-Host "Copying current workspace to $finalWorkingDirectory"
+        Copy-Item -Path $tmpSource -Destination $finalWorkingDirectory -Recurse -Force -ErrorAction SilentlyContinue| Out-Null
```
Comment (ader1990) [inline RunTests.ps1]: this is a vital operation and the erroraction should not be set to silently continue

**Comment by juhlee-microsoft** [comment]
I am in the same page with @ader1990. Please fix the code along with his comment and let me know. Will sign off then.

**Comment by iamshital** [comment]
@juhlee-microsoft I've addressed some issues mentioned by Adrian and created a new issue https://github.com/LIS/LISAv2/issues/93 and assigned it to myself. 
Remaining changes needs code evaluation and proper testing.

File: RunTests.ps1 (hunk 7 above)
Comment (juhlee-microsoft) [reply RunTests.ps1]: I see file copying and folder removal in 'finally' section. This should not remain unused code. I think this is ok.

File: Libraries/Framework.psm1 [hunk 9]
```diff
@@ -917,8 +917,8 @@ Function UploadTestResultToDatabase ($TestPlatform,$TestLocation,$TestCategory,$
 				$dbpassword = $XmlSecrets.secrets.DatabasePassword
 				$database = $XmlSecrets.secrets.DatabaseName
 				$dataTableName = "LISAv2Results"
-				$SQLQuery = "INSERT INTO $dataTableName (DateTimeUTC,TestPlatform,TestLocation,TestCategory,TestArea,TestName,TestResult,SubTestName,SubTestResult,TestTag,GuestDistro,KernelVersion,LISVersion,HostVersion,VMSize,Networking,ARMImage,OsVHD,LogFile,BuildURL) VALUES "
-				$SQLQuery += "('$DateTimeUTC','$TestPlatform','$TestLocation','$TestCategory','$TestArea','$TestName','$testResult','','','$TestTag','$GuestDistro','$KernelVersion','$LISVersion','$HostVersion','$VMSize','$Networking','$ARMImageName','$OsVHD','$UploadedURL', '$BuildURL'),"
+				$SQLQuery = "INSERT INTO $dataTableName (DateTimeUTC,TestPlatform,TestLocation,TestCategory,TestArea,TestName,TestResult,SubTestName,SubTestResult,ExecutionTag,GuestDistro,KernelVersion,LISVersion,HostVersion,VMSize,Networking,ARMImage,OsVHD,LogFile,BuildURL) VALUES "
```
[the remark this answers is not in the document]
Comment (juhlee-microsoft) [reply Libraries/Framework.psm1]: Only verified Field name fixed. It is ok for now. Others go to #93

**Comment by juhlee-microsoft** [comment]
Decided to merge to master branch in 7/19 Open Source Automation and Release team Meeting

**Pull request merged**